proxy/http.go

```diff
@@ -0,0 +1,196 @@
+	}
+}
+
+var errTemplate = template.Must(template.New("proxy-error").Parse(`<!DOCTYPE html>
+<html>
+	<head>
+	    <title>Error</title>
+		<style>
+			html {
+				height: 100%;
+				display: table;
+				margin: auto;
+			}
+			body {
+				height: 100%;
+				display: table-cell;
+				vertical-align: middle;
+				background-color: white;
+			}
+		</style>
+	</head>
+	<body>
+		<span style='font-size: 40px; color: black; font-family:Arial,Helvetica,sans-serif;'>{{.}}</span>
+	</body>
+</html>`))
+
+func handleHttpError(err error, w http.ResponseWriter, r *http.Request) {
+	var body []byte
+	var contentType string
+	var status = httpStatus(err)
+
+	// If the error was caused by an issue with the upstream connection, then we must
+	// ask the client to close the connection and create a new one. This is important
+	// because the http proxy handler isn't currently able to re-establish connections
+	// in response to them breaking. So if the client continues to use this connection,
+	// then they will continue to get 5xx errors due to the broken upstream connection.
+	// Note that, while the `Connection` header is not compatible with http2, the Go
+	// http2 package seems to handle this by removing the header and sending a GOAWAY.
+	// In fact, this is the _only_ way I can figure out how to send a GOAWAY from a handler.
+	if status >= 500 {
+		w.Header().Add("Connection", "close")
+	}
+
+	var headers = r.Header["Accept"]
+	var accept string
+	if len(headers) > 0 {
+		accept = headers[0]
+	}
+	if strings.Contains(accept, "json") {
+		body, _ = json.Marshal(map[string]interface{}{
+			"error": err.Error(),
+		})
+		contentType = "application/json"
+	} else if strings.Contains(accept, "html") {
+		var buf bytes.Buffer
+		if templateErr := errTemplate.Execute(&buf, err.Error()); templateErr != nil {
+			// TODO:
+			log.WithFields(log.Fields{
+				"origError":     err.Error(),
+				"templateError": templateErr.Error(),
+			}).Error("error rendering html error template")
+			panic(fmt.Sprintf("Error rendering html error template: %v", templateErr))
+		}
+		body = buf.Bytes()
+		contentType = "text/html"
+	} else {
+		// just render as plain text
+		body = []byte(fmt.Sprintf("Error: %s", err))
+		contentType = "text/plain"
+	}
+
+	w.Header().Add("Content-Type", contentType)
+	w.Header().Add("Content-Length", strconv.Itoa(len(body)))
+	w.WriteHeader(status)
+	var _, writeErr = w.Write(body)
+	if writeErr != nil {
+		log.WithFields(log.Fields{
+			"origError":     err.Error(),
+			"templateError": writeErr.Error(),
+		}).Warn("failed to write error response body")
+	}
+}
+
+func httpStatus(err error) int {
+	if err == NoMatchingShard {
+		return 404
+	} else if err == auth.InvalidAuthHeader || err == InvalidAcceptHeader {
+		return 400
+	} else if err == auth.MissingAuthHeader {
+		return 401
+	} else if err == auth.Unauthorized {
+		// In this case, the user provided a valid auth token, which just didn't authorize them to access the shard. We return a 403
```

Comment:
There might be an argument to be made here that getting a valid auth token is trivial, and so returning 403 here opens a vector for somebody to use the API for maliciously determining if an endpoint is valid or not. I can't say that concerns me too much at this point, so the usability benefit is probably the way to go.